test/discrete/test_fqf.py

```diff
@@ -0,0 +1,156 @@
+    parser.add_argument('--lr', type=float, default=3e-3)
+    parser.add_argument('--fraction-lr', type=float, default=2.5e-9)
+    parser.add_argument('--gamma', type=float, default=0.9)
+    parser.add_argument('--num-fractions', type=int, default=32)
+    parser.add_argument('--num-cosines', type=int, default=64)
+    parser.add_argument('--ent-coef', type=float, default=10.)
+    parser.add_argument('--n-step', type=int, default=3)
+    parser.add_argument('--target-update-freq', type=int, default=320)
+    parser.add_argument('--epoch', type=int, default=10)
+    parser.add_argument('--step-per-epoch', type=int, default=10000)
+    parser.add_argument('--step-per-collect', type=int, default=10)
+    parser.add_argument('--update-per-step', type=float, default=0.1)
+    parser.add_argument('--batch-size', type=int, default=64)
+    parser.add_argument('--hidden-sizes', type=int,
+                        nargs='*', default=[64, 64, 64])
+    parser.add_argument('--training-num', type=int, default=10)
+    parser.add_argument('--test-num', type=int, default=100)
+    parser.add_argument('--logdir', type=str, default='log')
+    parser.add_argument('--render', type=float, default=0.)
+    parser.add_argument('--prioritized-replay',
+                        action="store_true", default=False)
+    parser.add_argument('--alpha', type=float, default=0.6)
+    parser.add_argument('--beta', type=float, default=0.4)
+    parser.add_argument(
+        '--device', type=str,
+        default='cuda' if torch.cuda.is_available() else 'cpu')
+    args = parser.parse_known_args()[0]
+    return args
+
+
+def test_fqf(args=get_args()):
+    env = gym.make(args.task)
+    args.state_shape = env.observation_space.shape or env.observation_space.n
+    args.action_shape = env.action_space.shape or env.action_space.n
+    # train_envs = gym.make(args.task)
+    # you can also use tianshou.env.SubprocVectorEnv
+    train_envs = DummyVectorEnv(
+        [lambda: gym.make(args.task) for _ in range(args.training_num)])
+    # test_envs = gym.make(args.task)
+    test_envs = DummyVectorEnv(
+        [lambda: gym.make(args.task) for _ in range(args.test_num)])
+    # seed
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    train_envs.seed(args.seed)
+    test_envs.seed(args.seed)
+    # model
+    feature_net = Net(args.state_shape, args.hidden_sizes[-1],
+                      hidden_sizes=args.hidden_sizes[:-1], device=args.device,
+                      softmax=False)
+    net = FullQuantileFunction(
+        feature_net, args.action_shape, args.hidden_sizes,
+        num_fractions=args.num_fractions, num_cosines=args.num_cosines,
+        device=args.device
+    )
+    optim = torch.optim.Adam(
+        list(net.preprocess.parameters()) + list(net.last.parameters())
+        + list(net.embed_model.parameters()), lr=args.lr
+    )
+    fraction_optim = torch.optim.RMSprop(
+        net.propose_model.parameters(), lr=args.fraction_lr
+    )
```

Comment:
this part is kind of messy. I know you have tried your best to unify this API with previous IQN network but is it possible to simplify, e.g., avoid using `net.last` and `net.embed_model`?

Reply:
Good point. I made some changes and it looks better now.